Remove Delaunay triangulation and use simplex integration instead

shellshocked2003 opened the pull request:
This branch removes all dependency on `Delaunay` for finding the volume and centroid of a block. It also is no longer necessary to do any meshing of the block faces. Instead, all calculations are done using simplex integration. Some of the logic implemented in `BlockVTK` has been moved into `Block` so that all vertices are oriented counter clockwise - oriented vertices are required to perform simplex integration. 

This addresses the issues in #32.

jhkolb commented on src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala:
```diff
@@ -138,6 +138,116 @@ object Block {
+  private def ccwCompare(pointA: Array[Double], pointB: Array[Double],
+                         center: Array[Double]): Boolean = {
+    // Check that points are in the same x-y plane
+    if (math.abs(NumericUtils.roundToTolerance(pointA(2), 5) -
```
Is there a reason we can't use the default tolerance of 6 in `roundToTolerance`?

shellshocked2003 replied:
No, this is a leftover from before. I'll switch it to the default.

jhkolb commented on src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala:
```diff
@@ -138,6 +138,116 @@ object Block {
+    // Check that points are in the same x-y plane
+    if (math.abs(NumericUtils.roundToTolerance(pointA(2), 5) -
+      NumericUtils.roundToTolerance(pointB(2), 5)) > NumericUtils.EPSILON) {
+      throw new IllegalArgumentException("ERROR: Input to BlockVTK.ccwCompare: "+
```
Let's make this `.ccwCompare: " +` for consistent formatting.

jhkolb commented on src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala:
```diff
@@ -138,6 +138,116 @@ object Block {
+        }
+        // Order vertices in counter-clockwise orientation
+        val center = Block.findCenter(rotatedVerts)
+        val orderedVerts = {
```
I'm not certain, but I think you don't actually need the extra set of braces here. So you can write

``` scala
val orderedVerts = if (...) {
    ...
} else {
    ...
}
```

shellshocked2003 replied:
They are gone now.

jhkolb commented on src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala:
```diff
@@ -446,84 +556,26 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
-        }
-      }
-    ).toMap
+  def orientedVertices: Map[Face, Seq[Array[Double]]] = {
```
I know we discussed the issue of where to put this logic briefly yesterday.

I think we should keep the logic in the `orientedVertices` method in the `Block` class rather than putting it in the `Block` object. Does that sound okay to you?

shellshocked2003 replied:
Yeah, I agree. Will change this.

jhkolb commented on src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala:
```diff
@@ -532,41 +584,103 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
-        3.0 * centroidY / totalVolume,
-        3.0 * centroidZ / totalVolume
-      )
+      val increments: Seq[(Double, Double, Double)] = faces.flatMap { face =>
```
Do we actually need to include the type explicitly here? Otherwise we can just write `val increments = faces.flatMap ...`

shellshocked2003 replied:
No, it's not needed. Leftovers from debugging, sorry...

jhkolb commented on src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala:
```diff
@@ -532,41 +584,103 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
-      )
+      val increments: Seq[(Double, Double, Double)] = faces.flatMap { face =>
+        vertices(face).map { faceVertex =>
+          val vertexIndex = vertices(face).indexOf(faceVertex)
```
I think you can avoid this `indexOf` call, which I could imagine getting a bit expensive, with something like this:

``` scala
vertices(face).zipWithIndex.map { case (index, faceVertex) =>
    ...
}
```

shellshocked2003 replied:
I don't think this is needed anymore after switching to `sliding`. That was to make sure that the map "stops" once it gets to the third last vertex. Ugly code, I know.

jhkolb commented on src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala:
```diff
@@ -532,41 +584,103 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
+
+          // Tetrahedra are constructed using three vertices at a time, so stop
+          // mapping beyond 3 to last vertex
+          if ( vertexIndex < vertices(face).size - 2 ) {
```
Ah, so is the whole index thing just meant to omit the last 3 elements? Why not just do `vertices(face).dropRight(3).map { ... }` instead?

shellshocked2003 replied:
Can't do `dropRight` since those elements are used in the calculations as well. I'll use `sliding` like we discussed.

jhkolb commented on src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala:
```diff
@@ -532,41 +584,103 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
+              vertices(face)(vertexIndex + 2)(1) - centerY,
+              vertices(face)(vertexIndex + 2)(2) - centerZ)
+
+            // Initialize Jacobian
```
I think you can initialize the Jacobian a little more cleanly. Probably you can write something like

``` scala
val Jacobian = DenseMatrix[Double](vertex1, vertex2, vertex3)
```

You may have to look at some of the Breeze documentation.

shellshocked2003 replied:
Yeah, I'll take care of this.

jhkolb commented on src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala:
```diff
@@ -532,41 +584,103 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
-
-        val n = linalg.cross(b - a, c - a)
-        a dot n
+      vertices(face).map { faceVertex =>
```
The comments from the `centroid` method may also apply here: initializing the Jacobian more cleanly and using `dropRight` to avoid the complex logic based on indexes.

jhkolb commented on src/test/scala/edu/berkeley/ce/sparkrocks/BlockSpec.scala:
```diff
@@ -429,6 +412,9 @@ class BlockSpec extends FunSuite {
     val face6 = Face(Array(0.0, 0.0, -1.0), 1.0, phi=0, cohesion=0)
     val block = Block(Array(0.0, 0.0, 0.0), List(face1, face2, face3, face4, face5, face6))
     val centroid = block.centroid
+    println("THIS IS THE CENTROID:")
```
Let's remove these print statements.

shellshocked2003 replied:
Sorry, leftovers.

jhkolb commented on src/test/scala/edu/berkeley/ce/sparkrocks/BlockSpec.scala:
```diff
@@ -443,6 +429,9 @@ class BlockSpec extends FunSuite {
     val block = Block(Array(0.5, 0.5, 0.5), List(face1, face2, face3, face4, face5, face6))
     val centroid = block.centroid
     val expectedCentroid = Array(0.5, 0.5, 1.0)
+    println("THIS IS THE CENTROID:")
```
Let's remove these print statements.

jhkolb commented:
Finished a first pass. A few small things to consider. Looks like you've really simplified some of the volume and centroid calculations!

shellshocked2003 commented:
Addressed all your comments. It's a lot cleaner now - used `sliding` for all the simplex integration stuff.

jhkolb commented on src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala:
```diff
@@ -446,84 +515,54 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
 
     // Check if block is extremely small
-    if (totalVolume <= NumericUtils.EPSILON) {
+    if (volume <= NumericUtils.EPSILON) {
```
Let's clean up the logic a bit here. I think you can write something like `allVertices.map(_(0)).sum` instead of `allVertices.map { v => v(0) }` and similarly for the other sums.

shellshocked2003 replied:
Done

jhkolb commented on src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala:
```diff
@@ -446,84 +515,54 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
-      faces.map { face =>
+  def orientedVertices: Map[Face, Seq[Array[Double]]] = {
+    val faceVertices = calcVertices
+    faceVertices.keys.zip(
```
Now that I think about it, you should be able to express this in a more idiomatic way. I believe you can just call `map` (or better yet `transform`) on the `Map` instance you get from `calcVertices` instead of using `keys.zip` and `toMap`.

shellshocked2003 replied:
That's actually pretty nice, much cleaner.

jhkolb commented on src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala:
```diff
@@ -532,41 +571,70 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
-        3.0 * centroidY / totalVolume,
-        3.0 * centroidZ / totalVolume
-      )
+      val increments = faces.flatMap { face =>
```
How about just doing `vertices.flatMap { case (face, faceVertices) => ... }` to simplify things and get rid of the annoying `vertices(face)` usages? Also, can we rename `vertices` to something that better reflects the fact that it's a `Map` and not a `Seq`?

shellshocked2003 replied:
Done

jhkolb commented on src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala:
```diff
@@ -532,41 +571,70 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
+          ( (vertex1(0) + vertex2(0) + vertex3(0)) * JacobianDet,
+            (vertex1(1) + vertex2(1) + vertex3(1)) * JacobianDet,
+            (vertex1(2) + vertex2(2) + vertex3(2)) * JacobianDet
+            )
```
Very nit picky, but I think this parenthesis isn't lined up.

shellshocked2003 replied:
That's how IntelliJ likes it. I've moved it up to the line above.

jhkolb commented on src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala:
```diff
@@ -532,41 +571,70 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
   private def findVolume: Double = {
-    val vertices = findVertices
-    val mesh = meshFaces(vertices)
+    val vertices = orientedVertices
```
I think this is similar to the above comments -- we can use something like `flatMap` so that we have more conventional Scala code.

shellshocked2003 replied:
Done!

jhkolb commented on src/main/scala/edu/berkeley/ce/sparkrocks/RockSlicer.scala:
```diff
@@ -14,7 +14,7 @@ object RockSlicer {
       .set("spark.kryo.registrationRequired", "true")
       // Register classes with Kryo using ClassRegistrator
-      .set("spark.kryo.registrator", "edu.berkeley.ce.rockslicing.ClassRegistrator")
+      .set("spark.kryo.registrator", "edu.berkeley.ce.sparkrocks.ClassRegistrator")
```
Nice!

jhkolb commented on src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala:
```diff
@@ -532,41 +571,70 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
-        val n = linalg.cross(b - a, c - a)
-        a dot n
+      // Tetrahedra are constructed using three vertices at a time
+      vertices(face).sliding(3).map { vertexTuples =>
```
How about doing `map { case (vertex1, vertex2, vertex3) => ... }` to avoid writing `vertexTuples(0)` and so on? Do you think that would simplify the code?

shellshocked2003 replied:
Updated this. `vertexTuples` was a bad name since it's actually a `Seq`...

jhkolb commented on src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala:
```diff
@@ -532,41 +571,70 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
+        val anchorVertex = vertices(face)(0)
+
+        // Tetrahedra are constructed using three vertices at a time
+        vertices(face).takeRight(vertices(face).length - 1).sliding(2).map { faceVertices =>
```
I think you can simplify this to a much cleaner `drop(1)`, and if you want you can use `.head` instead of `(0)`  for the anchor vertex. :)

shellshocked2003 replied:
Yeah, don't know why I didn't just write that the first time round. I think leaving it as `(0)` is better for consistency since there's a bunch of `(1)` and `(2)` in there too, if that makes sense.

jhkolb commented:
A few more small things, but I think they'll go a long way towards making our code more conventional so other Scala programmers can more easily understand what's going on.

I never seem to find everything on the first pass.

shellshocked2003 commented:
Addressed all your comments. I think it's inevitable that you'll miss some stuff on the first pass. Some of the code was just moved from `BlockVTK` to `Block` though, so I guess this makes it the fourth pass ;)

jhkolb commented:
Looks good. I'll merge it.
